### PR TITLE
gh-93678: apply remove_redundant_jumps in optimize_cfg

### LINKED ISSUE #93678: Direct unit tests for compiler optimisations
We currently do not have a way to directly test compiler optimisations, and the tests we have are not always reliable ([more background](https://github.com/faster-cpython/ideas/issues/405)).

We need to be able to feed a sequence of bytecode instructions into the optimiser from python, and get its output back into python. Then we can write tests.

### PR DESCRIPTION
**Pull request by iritkatriel**
Fixes #93678.

Apply remove_redundant_jumps in optimise_cfg because it belongs there. Later in move_cold_blocks_to_end, we may need to apply it again. 

Added an assertion just before assembly stage that there are no redundant jumps.

Also tidy up an unrelated error handling issue I spotted.

* Issue: gh-93678

### CODE REVIEW

File: Lib/test/test_peepholer.py [hunk 1]
```diff
@@ -903,8 +904,7 @@ def test_conditional_jump_forward_const_condition(self):
         ]
         expected = [
             ('NOP', None, 11),
-            ('JUMP', lbl := self.Label(), 12),
-            lbl,
+            ('NOP', None, 12),
             ('LOAD_CONST', '3', 14)
```
Comment (mdboom) [inline Lib/test/test_peepholer.py:908]: It's cool to see the effect of this in the test change.

File: Python/compile.c [hunk 2]
```diff
@@ -8476,6 +8482,19 @@ propagate_line_numbers(basicblock *entryblock);
 static void
 eliminate_empty_basic_blocks(cfg_builder *g);
 
+#ifdef Py_DEBUG
+static void
+assert_no_redundant_jumps(cfg_builder *g) {
```
Comment (markshannon) [inline Python/compile.c]: Convert this to `int no_redundant_jumps(cfg_builder *g)` so it can be asserted.

File: Python/compile.c [hunk 3]
```diff
@@ -8627,6 +8642,10 @@ assemble(struct compiler *c, int addNone)
         goto error;
     }
 
+#ifdef Py_DEBUG
+    assert_no_redundant_jumps(g);
```
Comment (markshannon) [inline Python/compile.c]: ... then this becomes `assert(no_redundant_jumps())` and you can drop the `#ifdef Py_DEBUG`

**Review by markshannon — COMMENTED** [review]
One minor refactoring suggestion.
Otherwise, looks good.

**Comment by iritkatriel** [comment]
> One minor refactoring suggestion.
> Otherwise, looks good.

So the way it is now, the Gentoo Non-Debug build fails because no_redundant_jumps is not defined:

```
Python/compile.c:8647:12: error: implicit declaration of function ‘no_redundant_jumps’; did you mean ‘remove_redundant_jumps’? [-Werror=implicit-function-declaration]
 8647 |     assert(no_redundant_jumps(g));
      |            ^~~~~~~~~~~~~~~~~~
```

If I remove the ``#ifdef Py_DEBUG`` from the definition of no_redundant_jumps then I get warnings about it not being used on my local build.

**Comment by iritkatriel** [comment]
@markshannon how about this version?

**Comment by iritkatriel** [comment]
> @markshannon how about this version?

Buildbots are fine with it.

**Comment by markshannon** [comment]
I'd rather fix the underlying issue.
Why does the Gentoo Non-Debug build have asserts on? That doesn't sound like non-debug to me.

**Comment by markshannon** [comment]
Changing the `#ifdef Py_DEBUG` to `#ifndef NDEBUG` should work, but I thought `#ifdef Py_DEBUG` implied `#ifndef NDEBUG`  ande vice-versa.

Am I wrong, or the gentoo non-debug build mis-configured somehow?

**Comment by gpshead** [comment]
`NDEBUG` can always be used independent of `Py_DEBUG`.  _(ex: we regularly build Python and the world `-UNDEBUG` mode at work in the default build that people test against)_

**Comment by gpshead** [comment]
as for how it happened in this case, CFLAGS adds -UNDEBUG in https://github.com/python/buildmaster-config/blob/0669211daf34f86e879868364efe15683da0e069/master/custom/factories.py#L277

**Review by markshannon — APPROVED** [review]

**Pull request merged**